protocol/x/vault/keeper/withdraw.go

```diff
@@ -0,0 +1,117 @@
+//   - average_skew = integral / (posterior_leverage - leverage)
+//   - integral = skew_antiderivative(skew_factor, posterior_leverage) -
+//     skew_antiderivative(skew_factor, leverage)
+//   - posterior_leverage = leverage / (1 - withdrawal_portion)
+//     = leverage / (1 - shares_to_withdraw / total_shares)
+//     = leverage * total_shares / (total_shares - shares_to_withdraw)
+func (k Keeper) GetVaultWithdrawalSlippagePpm(
+	ctx sdk.Context,
+	vaultId types.VaultId,
+	sharesToWithdraw *big.Int,
+) (*big.Int, error) {
+	totalShares := k.GetTotalShares(ctx).NumShares.BigInt()
+	if sharesToWithdraw.Sign() <= 0 || sharesToWithdraw.Cmp(totalShares) > 0 {
+		return nil, errorsmod.Wrapf(
+			types.ErrInvalidSharesToWithdraw,
+			"sharesToWithdraw: %s, totalShares: %s",
+			sharesToWithdraw,
+			totalShares,
+		)
+	}
+
+	quotingParams, exists := k.GetVaultQuotingParams(ctx, vaultId)
+	if !exists {
+		return nil, types.ErrVaultParamsNotFound
+	}
+
+	_, perpetual, marketParam, marketPrice, err := k.GetVaultClobPerpAndMarket(ctx, vaultId)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get vault leverage.
+	leveragePpm, _, err := k.GetVaultLeverageAndEquity(ctx, vaultId, perpetual, marketPrice)
+	if err != nil {
+		return nil, err
+	}
+	// No leverage, no slippage.
+	if leveragePpm.Sign() == 0 {
+		return big.NewInt(0), nil
+	}
+
+	// Use absolute value of leverage.
+	leveragePpm.Abs(leveragePpm)
+
+	// Calculate simple_slippage = leverage * initial_margin (round up if necessary).
+	lt, err := k.perpetualsKeeper.GetLiquidityTier(ctx, perpetual.Params.LiquidityTier)
+	if err != nil {
+		return nil, err
+	}
+	bigOneMillion := lib.BigIntOneMillion()
+	simpleSlippagePpm := new(big.Int).Mul(
+		leveragePpm,
+		new(big.Int).SetUint64(uint64(lt.InitialMarginPpm)),
+	)
+	simpleSlippagePpm = lib.BigDivCeil(simpleSlippagePpm, bigOneMillion)
+
+	// Return simple slippage if withdrawing 100%.
+	if sharesToWithdraw.Cmp(totalShares) == 0 {
+		return simpleSlippagePpm, nil
+	}
+
+	// Estimate slippage.
+	// 1. leverage_after_withdrawal
+	//    = leverage / (1 - withdrawal_portion)
+	//    = leverage * total_shares / (total_shares - shares_to_withdraw)
+	posteriorLeveragePpm := new(big.Int).Mul(leveragePpm, totalShares)
+	posteriorLeveragePpm = lib.BigDivCeil(
+		posteriorLeveragePpm,
+		new(big.Int).Sub(totalShares, sharesToWithdraw),
+	)
+
+	// 2. integral = skew_antiderivative(skew_factor, posterior_leverage) - skew_antiderivative(skew_factor, leverage)
+	estimatedSlippagePpm := vault.SkewAntiderivativePpm(quotingParams.SkewFactorPpm, posteriorLeveragePpm)
+	estimatedSlippagePpm.Sub(estimatedSlippagePpm, vault.SkewAntiderivativePpm(quotingParams.SkewFactorPpm, leveragePpm))
+
+	// 3. average_skew = integral / (posterior_leverage - leverage)
+	estimatedSlippagePpm.Mul(estimatedSlippagePpm, bigOneMillion)
+	estimatedSlippagePpm = lib.BigDivCeil(
+		estimatedSlippagePpm,
+		posteriorLeveragePpm.Sub(posteriorLeveragePpm, leveragePpm),
+	)
```

Comment:
Since `posterior_leverage` and `leverage` are often close, as withdrawals get smaller, this denominator loses precision.

Would it be better to treat both leverages as ratios?

Define
- `o` open notional (unsigned)
- `e` vault equity, such that initial leverage `l0 = o / e`
- `m` the integer number of shares being withdrawn
- `n` the integer number of total shares

Then define posterior leverage `l1 = l0 / (1 - m/n) = (o/e) / (1 - m/n)`

And define change in leverage `dl` (which is the denominator here):

    dl = l1 - l0 = om / e(n - m)

Now average skew s is still `s = integral / dl`, but since we have `dl` as a ratio, you can invert it and take:

    s = integral / dl = integral * e(n - m) / om

I think this avoids the loss of precision discussed [here](https://github.com/dydxprotocol/v4-chain/pull/2242/files/78ded0f572e3dfa81044943ff14d769401b86abb#r1757872808)